code/Tree.cs

```diff
@@ -54,7 +130,7 @@ private int NewId()
             {
                 id++;
             }
-            while (children.ContainsKey(id));
+            while (parent.ContainsKey(id));
```

Comment:
@imane-aanna 
The complexity is O(n).
You can also use something like:
 max(parent)+1